File: cpp/src/grpc/adapter.cpp
#include <iostream>
#include "adapter.h"
#include "../until/timer.h"
#include "../service/observer.h"
#include "../service/provider.h"

GrpcWorker::GrpcWorker()
{
    cout << "GrpcWorker" << endl;
}

GrpcWorker::~GrpcWorker()
{
    cout << "~GrpcWorker" << endl;
}

GrpcAdapter::GrpcAdapter(GrpcAdaptee *p)
{
    cout << "GrpcAdapter" << endl;
    adaptee = p;
    string host = p->getGrpcHost();
    if (host.empty())
    {
        host = "localhost:5000";
    }
    std::shared_ptr<Channel> channel = grpc::CreateChannel(host, grpc::InsecureChannelCredentials());
    stub = zdautomotive::protobuf::ServiceManager::NewStub(channel);
    isWaitForWorkAgain = false;
}

GrpcAdapter::~GrpcAdapter()
{
    cout << "~GrpcAdapter" << endl;
}

void GrpcAdapter::doWork()
{
    // sendPingReqToGrpcServer();
    // sendRegisterReqToGrpcServer();
    // sendKeepAliveReqToGrpcServer();
    // waitForBindSubscriberCompleted();
}

void GrpcAdapter::stopWork()
{
    cout << "observer stop work" << endl;
}

void GrpcAdapter::waitForWorkAgain()
{
    if (isWaitForWorkAgain)
    {
        cout << "wait For do Work again" << endl;
        Timer t;
        t.setTimeout([&]()
                     { doWork(); },
                     10 * 1000);
    }
}

void GrpcAdapter::sendPingReqToGrpcServer()
{
    cout << "send ping req to grpc server" << endl;
    ClientContext context;
    Empty req;
    Empty res;
    Status status = stub->ping(&context, req, &res);
    if (!status.ok())
    {
        cout << "grpc server is offline." << endl;
        waitForWorkAgain();
    }
    else
    {
        cout << "grpc server is online." << endl;
    }
}

void GrpcAdapter::sendRegisterReqToGrpcServer()
{
    cout << "send register req to grpc server" << endl;
    ClientContext context;
    ZDServiceSubscriber subscriber;
    string subscriberId = adaptee->getSubscriberId();
    cout << "subscriberId:" << subscriberId << endl;
    subscriber.set_id(subscriberId);
    ZDServiceSubscriber_SUBSCRIBER_TYPE subscriberType = adaptee->getSubscriberType();
    cout << "subscriberType:" << subscriberType << endl;
    subscriber.set_type(subscriberType);
    SERVICE_TYPE *p_service_type = adaptee->getSubscriberServiceTypes();
    int sub_service_num = adaptee->getSubscriberServiceTypesNum();
    for (int i = 0; i < sub_service_num; i++)
    {
        subscriber.add_servicetype(p_service_type[i]);
        cout << "add_servicetype:" << p_service_type[i] << endl;
        p_service_type++;
    }
    ZDServiceRequest request;
    unique_ptr<ClientReader<ZDServiceRequest>> reader(stub->subscribe(&context, subscriber));
    while (reader->Read(&request))
    {
        if (adaptee != NULL)
        {
            cout << "handle req by subscriber type:" << subscriberType << endl;
            if (subscriberType == ZDServiceSubscriber_SUBSCRIBER_TYPE::ZDServiceSubscriber_SUBSCRIBER_TYPE_PROVIDER)
            {
                ZDServiceRequestResult result = ((Provider *)adaptee)->onSubscriberServiceRequest(request);
                submitRequestResult(result);
            }
            else
            {
                ((Observer *)adaptee)->onSubscriberServiceRequest(request);
            }
        }
    }
    // Status status = reader->Finish();
    // if (status.ok())
    // {
    //     cout << "finish register stream succeeded." << endl;
    // }
}

void GrpcAdapter::submitRequestResult(ZDServiceRequestResult &result)
{
    ClientContext context;
    ZDResponse res;
    Status status = stub->submitRequestResult(&context, result, &res);
    if (!status.ok())
    {
        cout << "submit req grpc server failed." << endl;
    }
    else
    {
        cout << "submit req grpc server successed." << endl;
        cout << "code:" << res.code() << endl;
        cout << "message:" << res.message() << endl;
    }
}

void GrpcAdapter::sendKeepAliveReqToGrpcServer()
{
    cout << "send keep alive req to grpc server." << endl;
    ClientContext context;
    shared_ptr<ClientReaderWriter<ZDSubscriberStatus, ZDResponse>> stream(stub->keepAlive(&context));
    ZDSubscriberStatus subscriber_status;
    string subscriberId = adaptee->getSubscriberId();
    subscriber_status.set_id(subscriberId);
    ZDServiceSubscriber_SUBSCRIBER_TYPE subscriberType = adaptee->getSubscriberType();
    if (subscriberType == ZDServiceSubscriber_SUBSCRIBER_TYPE::ZDServiceSubscriber_SUBSCRIBER_TYPE_PROVIDER)
    {
        ZDService *p_service_status = ((Provider *)adaptee)->getCurrentServiceStatus();
        int sub_service_num = adaptee->getSubscriberServiceTypesNum();
        for (int i = 0; i < sub_service_num; i++)
        {
            ZDService *p_service = subscriber_status.add_service();
            SERVICE_TYPE service_type = p_service_status->type();
            ZDService_SERVICE_STATUS service_status = p_service_status->status();
            p_service->set_type(service_type);
            p_service->set_status(service_status);
            p_service_status++;
            cout << "add_service:" << service_type << "," << service_status << endl;
        }
    }
    stream->Write(subscriber_status);
    stream->WritesDone();
    std::this_thread::sleep_for(std::chrono::milliseconds(5000));

    ZDResponse res;
    while (stream->Read(&res))
    {
        cout << "get response from keep alive stream ... " << endl;
        cout << "code:" << res.code() << endl;
        cout << "message:" << res.message() << endl;
    }
    // Status status = stream->Finish();
    // if (!status.ok())
    // {
    //     cout << "finish keep alive stream failed." << endl;
    // }
}

void GrpcAdapter::waitForBindSubscriberCompleted()
{
    cout << "waitForBindSubscriberCompleted" << endl;
    // Timer t = new Timer();
    // t.setInterval([&]() { }, 10 * 1000);
    // t.clearInterval();
}

File: node/src/lib/grpc/adapter.ts
import * as grpc from '@grpc/grpc-js';
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
import { StringValue } from 'google-protobuf/google/protobuf/wrappers_pb';
import { ZDServiceSubscriber, ZDSubscriberStatus, ZDServiceRequestResult } from '../proto/manager_pb';
import { ServiceManagerClient } from '../proto/manager_grpc_pb';
import { ZDResponse } from '../proto/wrappers_pb';
import { GrpcProviderAdaptee, GrpcSubscriberAdaptee } from './adaptee';
import moment from 'moment';
import logger from '../utils/logger';

const TIME_KEEP_INTERAL = 60 * 1000;
const TIME_RECONNECT_INTERAL = 10 * 1000;
const TIME_DELAY_BIND = 15 * 1000;

export abstract class GrpcWorker {
    abstract doWork(): void;
    abstract stopWork(): void;
}

export class GrpcAdapter extends GrpcWorker {

    private adaptee: GrpcProviderAdaptee | GrpcSubscriberAdaptee;
    private client: ServiceManagerClient;
    private keepAliveStream: grpc.ClientDuplexStream<ZDSubscriberStatus, ZDResponse> | undefined = undefined;
    private keepAliveHandler: NodeJS.Timeout | null = null;
    private isWaitForWorkAgain = false;

    constructor(adaptee: GrpcProviderAdaptee | GrpcSubscriberAdaptee) {
        super();
        this.adaptee = adaptee;
        const host: string = this.adaptee.getGrpcHost();
        this.client = new ServiceManagerClient(host, grpc.credentials.createInsecure());
    }

    public doWork(): void {
        this.isWaitForWorkAgain = false;
        this.sendPingReqToGrpcServer()
            .then(() => {
                return this.sendRegisterReqToGrpcServer();
            }).then(() => {
                return this.sendKeepAliveReqToGrpcServer();
            }).then((stream) => {
                this.waitForBindSubscriberCompleted(stream);
            }).catch((error: Error) => {
                logger.debug('doWork err:' + error.message + moment().format('MMMM Do YYYY, h:mm:ss a'));
                this.waitForWorkAgain();
            })

    }

    public stopWork(): void {
        if (this.keepAliveHandler) {
            clearInterval(this.keepAliveHandler);
            this.keepAliveHandler = null;
        }
        if (this.client) {
            this.client.unsubscribe(new StringValue().setValue(this.adaptee.getSubscriberId()), (err, res) => {
                if (!err && res) {
                    logger.info(res.getMessage());
                }
            })
        }
    }

    public sendSubscriberStatusToGrpcServer(): void {
        if (this.keepAliveStream) {
            if (this.isGrpcProviderAdapter(this.adaptee)) {
                const status: ZDSubscriberStatus = new ZDSubscriberStatus();
                const servicesList = this.adaptee.getCurrentServiceStatus();
                status.setId(this.adaptee.getSubscriberId());
                status.setServiceList(servicesList);
                this.keepAliveStream.write(status);
                logger.debug(`send provider status (id:${this.adaptee.getSubscriberId()},service:${servicesList.toString()}) to server . ${moment().format('MMMM Do YYYY, h:mm:ss a')}.`);
            } else {
                const status: ZDSubscriberStatus = new ZDSubscriberStatus();
                status.setId(this.adaptee.getSubscriberId());
                this.keepAliveStream.write(status);
                logger.debug(`send subscriber status (id:${this.adaptee.getSubscriberId()}) to server .${moment().format('MMMM Do YYYY, h:mm:ss a')}.`);
            }
        }
    }

    private waitForWorkAgain(): void {
        if (this.keepAliveHandler !== null) {
            clearInterval(this.keepAliveHandler);
            this.keepAliveHandler = null;
            logger.info('clear keep alive timer.');
        }
        if (!this.isWaitForWorkAgain) {
            this.isWaitForWorkAgain = true;
            setTimeout(() => {
                this.doWork();
            }, TIME_RECONNECT_INTERAL);
        }
    }

    private onReceiveGrpcServerKeepAliveReply(): void {
        if (this.keepAliveHandler === null) {
            this.keepAliveHandler = setInterval(() => {
                this.sendSubscriberStatusToGrpcServer();
            }, TIME_KEEP_INTERAL);
        }
        logger.info('receive keep alive reply');
    }

    private sendPingReqToGrpcServer(): Promise<void> {
        logger.debug(`send ping req to ${this.adaptee.getGrpcHost()}`);
        return new Promise<void>((resolve, reject) => {
            this.client.ping(new Empty(), (error) => {
                if (error) {
                    reject(new Error('grpc server is offline.'));
                } else {
                    logger.debug('grpc server is online.');
                    resolve();
                }
            });
        })
    }

    private sendRegisterReqToGrpcServer(): Promise<void> {
        logger.debug(`send register req to ${this.adaptee.getGrpcHost()}`);
        return new Promise<void>((resolve, reject) => {
            if (this.client) {
                const sub = new ZDServiceSubscriber();
                sub.setId(this.adaptee.getSubscriberId());
                sub.setType(this.adaptee.subscriberType);
                sub.setServicetypeList(this.adaptee.getSubscriberServiceTypes());
                const stream = this.client.subscribe(sub);
                stream.on('data', (req) => {
                    if (this.isGrpcProviderAdapter(this.adaptee)) {
                        const result: ZDServiceRequestResult | null = this.adaptee.onSubscriberServiceRequest(req);
                        if (result !== null) {
                            this.client.submitRequestResult(result, (err, res?) => {
                                if (!err && res) {
                                    logger.info('submit req :' + res.getMessage());
                                } else {
                                    logger.info('submit req error .');
                                }
                            });
                        }
                    } else {
                        this.adaptee.onSubscriberServiceRequest(req);
                    }
                });
                stream.on('error', (err) => {
                    logger.debug('grpc client register stream error:', err.message);
                    this.waitForWorkAgain();
                })
                resolve();
            } else {
                reject(new Error('register error.'));
            }
        })
    }

    private sendKeepAliveReqToGrpcServer(): Promise<grpc.ClientDuplexStream<ZDSubscriberStatus, ZDResponse>> {
        logger.debug('send keep alive req to grpc server.');
        return new Promise<grpc.ClientDuplexStream<ZDSubscriberStatus, ZDResponse>>((resolve, reject) => {
            if (this.client) {
                const stream = this.client.keepAlive();
                if (stream) {
                    this.keepAliveStream = stream;
                    if (this.keepAliveHandler === null) {
                        // setTimeout(() => {
                        this.sendSubscriberStatusToGrpcServer();
                        // }, TIME_DELAY_BIND);
                    }
                    resolve(stream);
                } else {
                    reject(new Error('keep alive error.'));
                }
            }
        })
    }

    private waitForBindSubscriberCompleted(stream: grpc.ClientDuplexStream<ZDSubscriberStatus, ZDResponse>): void {
        logger.debug('wait grpc server bind subscriber completed.');
        stream.on('data', (response: ZDResponse) => {
            // logger.info('recive pong from grpc server', response.getCode());
            if (response.getCode() === ZDResponse.ERROR_CODE.OK) {
                this.onReceiveGrpcServerKeepAliveReply();
            } else {
                logger.debug('grpc client keep alive bind error:' + response.getMessage());
                this.waitForWorkAgain();
            }
        });
        stream.on('error', (error) => {
            logger.debug('grpc client keep alive stream error.', error.message);
            this.waitForWorkAgain();
        });
        stream.on('end', () => {
            logger.debug('grpc client keep alive stream end.');
            this.waitForWorkAgain();
        });
    }

    private isGrpcProviderAdapter(adapter: GrpcProviderAdaptee | GrpcSubscriberAdaptee): adapter is GrpcProviderAdaptee {
        return adapter.subscriberType === ZDServiceSubscriber.SUBSCRIBER_TYPE.PROVIDER;
    }
}




